File: leetcode_problems/typescript/src/lc2670_find_distinct_diff_arr.ts
function distinctDifferenceArray(nums: number[]): number[] {
  let pd = 0;
  let sd = 0;
  let diff: number[] = [];
  let pm = new Map<number, number>();
  let sm = new Map<number, number>();

  for (let num of nums) {
    if ((pm.get(num) || 0) == 0) {
      pd++;
    }
    pm.set(num, (pm.get(num) || 0) + 1);
  }

  for (let num of nums) {
    pm.set(num, (pm.get(num) || 0) - 1);
    if ((sm.get(num) || 0) == 0) {
      sd++;
    }
    sm.set(num, (sm.get(num) || 0) + 1);
    if ((pm.get(num) || 0) == 0) {
      pd--;
    }
    diff.push(sd - pd);
  }

  return diff;
}

console.log(distinctDifferenceArray([1, 2, 3, 4, 5]));

File: leetcode_problems/cpp/lc2670_find_distinct_diff_arr.cpp
#include <vector>
#include <iostream>
#include <unordered_map>
using std::unordered_map;
using std::vector;

class Solution
{
public:
    vector<int> distinctDifferenceArray(vector<int> &nums)
    {
        int prefdist = 0;
        int suffdist = 0;
        vector<int> diff;
        unordered_map<int, int> prefmap;
        unordered_map<int, int> suffmap;

        for (int num : nums)
        {
            if (prefmap[num] == 0)
            {
                prefdist++;
            }
            prefmap[num]++;
        }

        for (int num : nums)
        {
            prefmap[num]--;
            if (suffmap[num] == 0)
            {
                suffdist++;
            }
            suffmap[num]++;
            if (prefmap[num] == 0)
            {
                prefdist--;
            }
            diff.push_back(suffdist - prefdist);
        }

        return diff;
    }
};

int main()
{
    Solution sol;
    vector<int> nums = {1, 2, 3, 4, 5};
    vector<int> result = sol.distinctDifferenceArray(nums);
    for (int val : result)
    {
        std::cout << val << " ";
    }
    std::cout << std::endl;
    return 0;
}